transport/tchannel/inbound.go

```diff
@@ -0,0 +1,64 @@
+package tchannel
+
+import (
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/introspection"
+	"go.uber.org/yarpc/internal/sync"
+)
+
+// Inbound receives YARPC requests over TChannel. It may be constructed using
+// the NewInbound method on a tchannel.Transport.
+type Inbound struct {
+	once      sync.LifecycleOnce
+	transport *Transport
+}
+
+// NewInbound returns a new TChannel inbound backed by a shared TChannel
+// transport.
+// There should only be one inbound for TChannel since all outbounds send the
+// listening port over non-ephemeral connections so a service can deduplicate
+// locally- and remotely-initiated persistent connections.
+func (t *Transport) NewInbound() *Inbound {
+	return &Inbound{
+		transport: t,
+	}
+}
+
+// SetRouter configures a router to handle incoming requests.
+// This satisfies the transport.Inbound interface, and would be called
+// by a dispatcher when it starts.
+func (i *Inbound) SetRouter(router transport.Router) {
+	i.transport.router = router
+}
+
+// Transports returns a slice containing the Inbound's underlying
+// Transport.
+func (i *Inbound) Transports() []transport.Transport {
+	return []transport.Transport{i.transport}
+}
+
+// Start starts this Inbound. Note that this does not start listening for
+// connections; that occurs when you start the underlying ChannelTransport is
+// started.
+func (i *Inbound) Start() error {
+	return i.once.Start(nil)
```

Comment:
we should make sure to prioritize https://github.com/yarpc/yarpc-go/issues/721 before we adopt this more widely